crates/uv-distribution/src/distribution_database.rs

```diff
@@ -577,46 +573,64 @@ impl<'a, Context: BuildContext> DistributionDatabase<'a, Context> {
-        // Run all providers.
-        let mut resolved_priorities = FxHashMap::default();
-        for (name, provider) in &variants_json.providers {
-            debug!("Querying provider {name}");
-            let config = {
-                async {
-                    // TODO(konsti): That's not spec compliant
-                    let backend_name = provider.plugin_api.clone().unwrap_or(name.clone());
-                    let builder = self
-                        .build_context
-                        .setup_variants(backend_name, provider, BuildOutput::Debug)
-                        .await?;
-                    Ok::<_, Error>(builder.query(&known_properties).await?)
-                }
-                .instrument(info_span!("query_variant_provider"))
-                .await?
-            };
-            trace!(
-                "Found namespace {} with configs {:?}",
-                config.namespace, config
-            );
-            resolved_priorities.insert(config.namespace.clone(), config);
-        }
+        let provider_outputs: FxHashMap<String, Arc<VariantProviderOutput>> =
+            futures::stream::iter(variants_json.providers.iter())
+                .map(|(name, provider)| {
+                    let known_properties = Arc::clone(&known_properties);
+                    async move {
+                        let config = if self.build_context.variants().register(provider.clone()) {
+                            debug!("Querying provider `{name}` for variants");
+
+                            // TODO(konsti): That's not spec compliant
+                            let backend_name = provider.plugin_api.clone().unwrap_or(name.clone());
+                            let builder = self
+                                .build_context
+                                .setup_variants(backend_name, provider, BuildOutput::Debug)
+                                .await?;
+                            let config = builder.query(&known_properties).await?;
+                            trace!(
+                                "Found namespace {} with configs {:?}",
+                                config.namespace, config
+                            );
+
+                            let config = Arc::new(config);
+                            self.build_context
+                                .variants()
+                                .done(provider.clone(), config.clone());
+                            config
+                        } else {
+                            debug!("Reading provider `{name}` from in-memory cache");
+                            self.build_context
+                                .variants()
+                                .wait(provider)
+                                .await
+                                .expect("missing value for registered task")
+                        };
+                        Ok::<_, Error>((config.namespace.clone(), config))
```

Comment:
We now cache the provider calls in-memory (not to-disk), and run them in parallel.